### ISSUE TITLE
v 7.5.9 FormStorage.GetFormFromCache(Guid id) doesn't actually cache the forms

### ISSUE THREAD

**Issue by JasonElkin** [umbraco/Umbraco.Forms.Issues#829]
We've found that Umbraco Forms is causing some performance issues on a site that has a large number of forms and a lot of traffic.

Whenever a page that includes a form is loaded, _all_ the forms in the site are fetched from the IFileSystem. Every time.

Reproduction
------------

### Bug summary

All forms are always fetched from the IFileSystem, on every page that includes a form.

### Specifics

The method in the call stack that's calling `IFileSystem.GetFIle(string path)` is `FormStorage.GetFormFromCache(Guid id)` which, presumably, should be caching forms rather than hitting the `IFileSystem` every time.

After decompiling, it looks like two things are happening:

1. `GetFormFromCache(Guid id)` is actually calling `BaseFileSystemStorage<Form>.GetAll()`, rather than fetching and caching a single form.
~~2. `GetFormFromCache(Guid id)` is caching the enumerator rather than the results.~~
Reading decompiled lambdas is tricky.

Because we're using the `Our.Umbraco.FileSystemProviders.Azure.AzureBlobFileSystem` from the `UmbracoFileSystemProviders.Azure` package, fetching a form is a non-trivial operation and the extra time it takes to fetch a form blob storage, multiplied by the number of forms in the site, adds about 8 seconds of overhead to our pageloads.

### Steps to reproduce

Create multiple forms.
Include a single form on a page.
Load the page and `IFileSystem.OpenFile(string path)` will be called for every form in the site.
Refresh the page and `IFileSystem.OpenFile(string path)` will be called again for every form in the site.

Easiest way to debug this is to create a custom `IFileSystem` that wraps an existing implementation and drop a break point on `IFileSystem.GetFIle(string path)`;

### Expected result

Page would only load the form for that specific page.
Subsequent loads should be cached/faster.

**Comment by JasonElkin** [comment]
This looks to also be the case in latest v8 when using the IFileSystem for storage (as opposed to the database).

**Comment by AndyButland** [comment]
Hi - thanks for reporting this, however I'm having trouble replicating.  When working with a test website running with the latest Forms V7 code, I'm finding this request is cached.  As you say, they all forms from disk are requested via this `GetFormFromCache` method, and then the one matching the provided is returned.  But the full list of forms is cached, for 10 minutes cache in the runtime cache.

There is a `DisableFormCaching` setting, so would be worth just checking that isn't set to `True` in your installation.

Otherwise would you mind please sharing your code and config for your _"custom IFileSystem that wraps an existing implementation"_, and I'll give it a go with that to see if I can see an issue with this cache not working as you have reported.  Thanks.

**Comment by JasonElkin** [comment]
Thanks for looking @AndyButland, I must have misread the decompiled lambda.

`DisableFormCaching` isn't set to true.

I have since subscribed to the `FormsCacheRefresher` and noticed that it fires whenever I load a page that includes a form. So it could be that the cache is working but is being cleared every time.

**Comment by callumbwhyte** [comment]
Hey @AndyButland,

`DisableFormCaching` is definitely set to false in config.

Here is [JsonCachingFileSystem.zip](https://github.com/umbraco/Umbraco.Forms.Issues/files/9245873/JsonCachingFileSystem.zip) - this is the solution we have used to wrap the forms file system and prevent the issue.

Attaching a breakpoint to our `GetFiles` method you should see this still gets hit every time, even though Umbraco Forms is supposedly caching that call....

The following is decompiled from the Forms source (sorry...)

```
public Form GetFormFromCache(Guid id)
{
    Form form = ApplicationContext.Current.ApplicationCache.RuntimeCache.GetCacheItem<IEnumerable<Form>>("Forms.FormStorage.All", new Func<IEnumerable<Form>>(((BaseFileStorage<Form>)this).GetAll), new TimeSpan?(TimeSpan.FromMinutes(10.0))).SingleOrDefault<Form>((Func<Form, bool>)(x => x.Id == id));
    return form != null ? form.Clone<Form>() : (Form)null;
}
```

Here `GetAll` returns an IEnumerable, rather than the resulting list, and so it's being not properly cached.

I can replicate this in code by calling `ApplicationContext.Current.ApplicationCache.RuntimeCache.GetCacheItem("Forms.FormStorage.All")` (i.e. with no add Func) - my breakpoint in our custom file system is still being hit every time.

If I clear the item from the cache, the below code is the same as in Forms but with a `ToList()` – my breakpoint is hit only once as expected:

```
ApplicationContext.Current.ApplicationCache.RuntimeCache.GetCacheItem("Forms.FormStorage.All", () => store.GetAll().ToList())
```

Hope this helps!

**Comment by AndyButland** [comment]
Thanks for the additional detail @callumbwhyte and @JasonElkin.  This is proving kind of frustrating... I can see the issue you are describing that makes sense, as does the fix you propose... but I'm still struggling to replicate the problem.

I probably should just add a `.ToList()` and be done with it!  But I would like to be sure I can see the issue, mainly so can also confirm that we have fixed it with the change.

Would you mind reviewing what I've done here please, just to see if you can see what I may have missed to replicate your setup?

- Umbraco 7 CMS (7.15.7) and Forms 7 (7.5.9)
- Installed the starter kit
- Installed `UmbracoFileSystemProviders.Azure` (1,2,0).
- Added the following file system, that just does a pass through to the underlying Azure provider:

```
using Our.Umbraco.FileSystemProviders.Azure;
using System;
using System.Collections.Generic;
using System.IO;
using Umbraco.Core;
using Umbraco.Core.Cache;
using Umbraco.Core.IO;

namespace TestSite3
{
    public class PassThroughFileSystem : IFileSystem
    {
        private static AzureBlobFileSystem _filesystem;

        private static readonly IRuntimeCacheProvider _runtimeCache = ApplicationContext.Current.ApplicationCache.RuntimeCache;

        public PassThroughFileSystem(string containerName, string rootUrl, string connectionString)
        {
            _filesystem = new AzureBlobFileSystem(containerName, rootUrl, connectionString);
        }

        public PassThroughFileSystem(string containerName, string rootUrl, string connectionString, string maxDays)
        {
            _filesystem = new AzureBlobFileSystem(containerName, rootUrl, connectionString, maxDays, "true", "false");
        }

        public PassThroughFileSystem(string containerName, string rootUrl, string connectionString, string maxDays, string useDefaultRoute)
        {
            _filesystem = new AzureBlobFileSystem(containerName, rootUrl, connectionString, maxDays, useDefaultRoute, "false");
        }

        public PassThroughFileSystem(string containerName, string rootUrl, string connectionString, string maxDays, string useDefaultRoute, string usePrivateContainer)
        {
            _filesystem = new AzureBlobFileSystem(containerName, rootUrl, connectionString, maxDays, useDefaultRoute, usePrivateContainer);
        }

        public PassThroughFileSystem(string alias)
        {
            _filesystem = new AzureBlobFileSystem(alias);
        }

        public IEnumerable<string> GetDirectories(string path) => _filesystem.GetDirectories(path);

        public void DeleteDirectory(string path) => _filesystem.DeleteDirectory(path);

        public void DeleteDirectory(string path, bool recursive) => _filesystem.DeleteDirectory(path, recursive);

        public bool DirectoryExists(string path) => _filesystem.DirectoryExists(path);

        public void AddFile(string path, Stream stream) => _filesystem.AddFile(path, stream);

        public void AddFile(string path, Stream stream, bool overrideIfExists) => _filesystem.AddFile(path, stream, overrideIfExists);

        public IEnumerable<string> GetFiles(string path) => _filesystem.GetFiles(path);

        public IEnumerable<string> GetFiles(string path, string filter) => _filesystem.GetFiles(path, filter);

        public Stream OpenFile(string path) => _filesystem.OpenFile(path);

        public void DeleteFile(string path) => _filesystem.DeleteFile(path);

        public bool FileExists(string path) => _filesystem.FileExists(path);

        public string GetRelativePath(string fullPathOrUrl) => _filesystem.GetRelativePath(fullPathOrUrl);

        public string GetFullPath(string path) => _filesystem.GetFullPath(path);

        public string GetUrl(string path) => _filesystem.GetUrl(path);

        public DateTimeOffset GetLastModified(string path) => _filesystem.GetLastModified(path);

        public DateTimeOffset GetCreated(string path) => _filesystem.GetCreated(path);
    }
}
```
- Running Azure Storage Emulator
- Configured with the following in `FileSystemProviders.config`:

```
  <Provider alias="forms" type="TestSite3.PassThroughFileSystem, TestSite3">
    <Parameters>
      <add key="containerName" value="form-data" />
      <add key="rootUrl" value="http://127.0.0.1:10000/devstoreaccount1/" />
      <add key="connectionString" value="UseDevelopmentStorage=true" />
    </Parameters>
  </Provider>
```
- Created a form and rendered it on the starter kit's home page grid and contact form
- I've then put break points on the following methods in the `PassThroughFileSystem`
    - `GetFiles(string path, string filter)`
    - `OpenFile(string path)`
- Started the site, and confirmed the breakpoints are hit,
- Refreshed the site pages and can see they aren't hit and the form is rendered.
- Waited 10 minutes, refreshed the site and see the breakpoints are hit again.

**Comment by AndyButland** [comment]
Did you ever get chance to look further at this @JasonElkin please?  I'm going to close now partly as I could replicate, and partly as we are no longer working on Forms 7 other than dealing with any security related issues, following the [Umbraco 7 end of life schedule](https://umbraco.com/products/knowledge-center/long-term-support-and-end-of-life/umbraco-7-end-of-life-eol/) (which took this version into the security-only phase from July 2021).

Of course if you do consider it still a problem, and it is found on a later version, please feel free to open a new issue indicating the versions affected, ideally along with some feedback on the behaviour I was seeing in the last comment.

**Comment by JasonElkin** [comment]
Hi @AndyButland, I put a workaround in place for us, and haven't had the chance to dive into this further - if you couldn't replicate then closing makes sense.
